cranelift/wasm/src/code_translator.rs

```diff
@@ -3579,50 +3581,258 @@ where
-        builder.ins().arm64_stg(tagged_ptr, iter_ptr);
+        // === Block definitions
+        // st2g_loop_condition_block(iter_ptr, st2g_loop_counter)
+        let st2g_loop_condition_block =
+            block_with_params(builder, [ValType::I64, ValType::I64], environ)?;
+
+        // TODO: does this really require iter_ptr and counter as parameters, since they are never modified in the condition
+        // st2g_loop_body_block(iter_ptr, st2g_loop_counter)
+        let st2g_loop_body_block =
+            block_with_params(builder, [ValType::I64, ValType::I64], environ)?;
+
+        // TODO: potential optimization: since this isn't required for all generated code, can we conditionally define this block only when needed?
+        // last_stgs_block()
+        let last_stgs_block = block_with_params(builder, [], environ)?;
+
+        // next_block()
+        let next_block = block_with_params(builder, [], environ)?;
+
+        let st2g_loop_counter = builder.ins().iconst(I64, 0);
+        builder
+            .ins()
+            .jump(st2g_loop_condition_block, &[iter_ptr, st2g_loop_counter]);
+
+        // === st2g_loop_condition_block(iter_ptr, st2g_loop_counter)
+        builder.switch_to_block(st2g_loop_condition_block);
+
+        let iter_ptr = builder.block_params(st2g_loop_condition_block)[0];
+        let st2g_loop_counter = builder.block_params(st2g_loop_condition_block)[1];
+
+        let cond = builder.ins().icmp_imm(
+            IntCC::UnsignedLessThan,
+            st2g_loop_counter,
+            num_required_st2g_loops_i64,
+        );
+        canonicalise_brif(
+            builder,
+            cond,
+            st2g_loop_body_block,
+            &[iter_ptr, st2g_loop_counter],
+            if num_required_last_stgs == 0 {
+                next_block
+            } else {
+                last_stgs_block
+            },
+            &[],
+        );
+
+        // === st2g_loop_body_block(iter_ptr, st2g_loop_counter)
+        builder.switch_to_block(st2g_loop_body_block);
+
+        for i in 0..LOOP_UNROLL_THRESHHOLD_I64 {
+            let cur_ptr = builder.ins().iadd_imm(iter_ptr, i * 32);
+            builder.ins().arm64_st2g(tagged_ptr, cur_ptr);
+        }
+
+        let iter_ptr = builder
+            .ins()
+            .iadd_imm(iter_ptr, LOOP_UNROLL_THRESHHOLD_I64 * 32);
+        let st2g_loop_counter = builder.ins().iadd_imm(st2g_loop_counter, 1);
+
+        builder
+            .ins()
+            .jump(st2g_loop_condition_block, &[iter_ptr, st2g_loop_counter]);
+
+        builder.seal_block(st2g_loop_body_block);
+        builder.seal_block(st2g_loop_condition_block);
+
+        // === last_stgs_block()
+        builder.switch_to_block(last_stgs_block);
+
+        // TODO: consider adding iter_ptr as a parameter (or calculate statically like done here)
+        // instead of statically calculating a huge offset to the original iter_ptr, might improve cache locality
```

Comment:
Any thoughts on this?

Reply:
Yes, makes sense to increment iter_ptr. I think this will not improve cache locality, (we're accessing memory sequentially in both ways), but we might run into a problem if the offset is too large to fit into the immediate for the `stg` instructions.

Reply:
Ok. The question that remains: should we pass `iter_ptr` as a parameter to `last_stgs_block`, which we can then use as is as it has already been incremented correctly in the st2g loops, or should we not pass `iter_ptr` as parameter and just add the same offset, that we usually passed on to `st2g`, directly onto iter_ptr at compile time? I prefer the first option, as we are already using the `iter_ptr` as an iterator and it would save the addition of the offset.